contracts/FundManagement.sol

```diff
@@ -0,0 +1,136 @@
+// SPDX-License-Identifier: GPL-3.0-only
+pragma solidity 0.8.4;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+contract FundManagement is Ownable {
+    struct Fund {
+        uint256 totalAmount;
+        address asset;
+        uint256 requestWithdrawTime;
+        uint256 pendingWithdrawAmount;
+    }
+
+    uint256 public requestWindow = 2 days;
+    mapping(address => mapping(uint256 => Fund)) public funds;
+    mapping(address => uint256) public numberOfFunds;
+    mapping(address => uint256) public locked;
+
+    event CreateFund(address indexed manager, address asset, uint256 amount);
+    event RequestFunds(address indexed to, uint256 amount);
+    event Claim(address indexed to, uint256 amount);
+    event Clawback(address account, uint256 fundNumber);
+
+    constructor() {}
+
+    function requestFunds(uint256 fundNumber, uint256 amount) external {
+        Fund storage fund = funds[msg.sender][fundNumber];
+        uint256 totalWithdrawableAmount = fund.pendingWithdrawAmount + amount;
+        require(
+            totalWithdrawableAmount <= fund.totalAmount, 
+            "Requested amount plus withdrawable amount greater than total allocated funds"
+        );
+        
+        fund.requestWithdrawTime = block.timestamp;
+        fund.pendingWithdrawAmount = totalWithdrawableAmount;
+
+        emit RequestFunds(msg.sender, amount);
+    }
+
+    function claim(uint256 fundNumber) external {
+        Fund storage fund = funds[msg.sender][fundNumber];
+        require(
+            fund.pendingWithdrawAmount > 0, 
+            "No withdrawable funds"
+        );
+        require(
+            fund.pendingWithdrawAmount <= fund.totalAmount, 
+            "Withdrawable amount greater than total allocated funds"
+        );
+        require(
+            block.timestamp >= fund.requestWithdrawTime + requestWindow, 
+            "Your funds are not withdrawable yet"
+        );
+
+        require(
+            IERC20(fund.asset).transfer(msg.sender, fund.pendingWithdrawAmount), 
+            "Transfer failed"
+        );
+
+        emit Claim(msg.sender, fund.pendingWithdrawAmount);
+
+        locked[fund.asset] = locked[fund.asset] - fund.pendingWithdrawAmount;
+        fund.totalAmount = fund.totalAmount - fund.pendingWithdrawAmount;
+        fund.pendingWithdrawAmount = 0;
```

Comment:
State changes should be made before external calls

Reply:
The reason i put that line last is because if i set pendingWithdrawAmount to 0 first it would affect the others. Now I could use a temp variable to stored that number and emit the event at the end. 

How important is it that I make the emit event after the state change? I though that if an error occurs the whole thing doesn't go through so no chance of emitted event and then not having state change right?

Reply:
> How important is it that I make the emit event after the state change?

[Very](https://blog.sigmaprime.io/solidity-security.html#re-prevention).